src/engine/plugins/tile.cpp

```diff
@@ -673,59 +647,32 @@ Status TilePlugin::HandleRequest(const std::shared_ptr<datafacade::BaseDataFacad
+                    auto forward_datasource_vector = facade->GetUncompressedForwardDatasources(edge.packed_geometry_id);
+                    auto reverse_datasource_vector = facade->GetUncompressedReverseDatasources(edge.packed_geometry_id);
+                    auto forward_weight = forward_weight_vector[edge.fwd_segment_position];
+                    auto reverse_weight = reverse_weight_vector[reverse_weight_vector.size() - edge.fwd_segment_position - 1];
+                    auto forward_datasource = forward_datasource_vector[edge.fwd_segment_position];
+                    auto reverse_datasource = reverse_datasource_vector[reverse_datasource_vector.size() - edge.fwd_segment_position - 1];
 
-                    forward_datasource_vector.clear();
-                    forward_datasource_vector =
-                        facade->GetUncompressedForwardDatasources(edge.packed_geometry_id);
-                    forward_datasource = forward_datasource_vector[edge.fwd_segment_position];
-
-                    reverse_weight_vector.clear();
-                    reverse_weight_vector =
-                        facade->GetUncompressedReverseWeights(edge.packed_geometry_id);
+                    use_line_value(reverse_weight);
+                    use_line_value(forward_weight);
 
                     std::string name = facade->GetNameForID(edge.name_id);
-
-                    forward_weight_vector = facade->GetUncompressedForwardWeights(edge.packed_geometry_id);
-                    forward_weight = forward_weight_vector[edge.fwd_segment_position];
-
-                    forward_datasource_vector = facade->GetUncompressedForwardDatasources(edge.packed_geometry_id);
-                    forward_datasource = forward_datasource_vector[edge.fwd_segment_position];
-
-                    reverse_weight_vector = facade->GetUncompressedReverseWeights(edge.packed_geometry_id);
-
-                    reverse_weight = reverse_weight_vector[reverse_weight_vector.size() -
-                                                           edge.fwd_segment_position - 1];
-
-                    reverse_datasource_vector = facade->GetUncompressedReverseDatasources(edge.packed_geometry_id);
-                    reverse_datasource =
-                        reverse_datasource_vector[reverse_datasource_vector.size() -
-                                                  edge.fwd_segment_position - 1];
-
-                    // Keep track of the highest datasource seen so that we don't write unnecessary
-                    // data to the layer attribute values
-                    max_datasource_id = std::max(max_datasource_id, forward_datasource);
-                    max_datasource_id = std::max(max_datasource_id, reverse_datasource);
+                    if (name_offsets.find(name) == name_offsets.end())
+                    {
+                        names.push_back(name);
+                        name_offsets[name] = names.size() - 1;
```

Comment:
Hm do we need the name further down here? If not I would re-order those two lines and `std::move` the already allocated name into the vector. (push back has an rvalue overload)